### ISSUE TITLE
Google Cloud Local SSD

### ISSUE THREAD

**Issue by jamespinkerton** [gc3-uzh-ch/elasticluster#526]
Hi. Is there a way to create an instance in google cloud with local SSD's? The application I have in mind is very throughput sensitive and would require the speed of a local SSD.

Thanks!

**Comment by riccardomurri** [comment]
Doesn't `boot_disk_type=pd-ssd` do?  (See: http://elasticluster.readthedocs.io/en/latest/configure.html#additional-optional-configuration-keys-for-google-cloud )

**Comment by jamespinkerton** [comment]
pd probably stands for persistent disk? I'm talking about local SSD's, which are not the same as persistent disks.

**Comment by riccardomurri** [comment]
I see. Then they're not currently supported.

Adding code to instanciate local SSD should be rather straightforward.

Although I read in the docs that they should be formatted and mounted separately -- this would require new actions in the playbooks and a way to pass the device names of these "extra disks".  On the other hand, this is quite easy to do if you're willing to go an extra setup step to run a custom Ansible playbook.

**Comment by xiaoyanh** [comment]
Hi Riccardo,
I would also really appreciate it if elasticluster might one day include this feature. (Or show an example of how to implement this ourselves.)  I am currently training deep networks on ImageNet dataset (~150GB) and having local SSD would greatly speed up this process. Thank you! -- XY @monajemi

**Comment by riccardomurri** [comment]
I have code to instanciate VMs with local SSDs on Google Cloud in
branch "issues/#526" on http://github.com/riccardomurri/elasticluster .

The trouble with passing this information on to playbooks is that the
name of the local SSD disks (as seen from Linux) is not completely
predictable. (And the `diskName` parameter is not allowed for local SSDs.)

With the ``SCSI`` interface, local SSDs are named `google-local-ssd-N`:

    riccardo.murri@su16g3-master001:~$ ls /dev/disk/by-id/
    google-local-ssd-0  google-persistent-disk-0        scsi-0Google_EphemeralDisk_local-ssd-0  scsi-0Google_PersistentDisk_persistent-disk-0
    google-local-ssd-1  google-persistent-disk-0-part1  scsi-0Google_EphemeralDisk_local-ssd-1  scsi-0Google_PersistentDisk_persistent-disk-0-part1

However, with the ``NVME`` interface (which is likely what most people
want since it's higher-perf), they're just named `nvme-nvme.***`:

    riccardo.murri@su16g2-master001:~$ ls /dev/disk/by-id/
    google-persistent-disk-0                                           nvme-nvme_card_nvme_card
    google-persistent-disk-0-part1                                     scsi-0Google_PersistentDisk_persistent-disk-0
    nvme-nvme.1ae0-6e766d655f6361726400-6e766d655f6361726400-00000001  scsi-0Google_PersistentDisk_persistent-disk-0-part1
    nvme-nvme.1ae0-6e766d655f6361726400-6e766d655f6361726400-00000002

**Comment by riccardomurri** [comment]
The following Ansible playbook will build a single RAID0 device out of all the available local SSDs and make the storage available in directory `/srv/google-local-ssd`.

At the moment, I have only tested the playbook on Ubuntu 16.04 ("xenial") VMs.

The playbook may be applied to any ElastiCluster cluster by running::

        elasticluster setup my-cluster-name /path/to/localssd.yml

or just add this to the cluster's `[setup]` section::

        ansible_extra_args = /path/to/localssd.yml

Here's the Ansible playbook

```yaml
---
#
# This playbook is for site-local customization to ElastiCluster's
# playbooks.  It runs *after* any other playbook distributed with
# ElastiCluster has gotten its chance to run.
#
# An empty playbook is checked into the Git repository.  If you make
# any local modifications, please run `git update-index
# --assume-unchanged after.yml` to avoid committing them accidentally
# into ElastiCluster's main branch.
#
- name: Format and mount Google local SSD disks
  tags:
    - after
    - local
    - local-ssd
    - google
  hosts: all

  tasks:
    - name: Discover SSD device names (SCSI)
      shell: |
        ls -1 /dev/disk/by-id/google-local-ssd-[0-9]
      register: localssd_scsi_devices
      ignore_errors: yes

    - name: No SCSI SSD devices
      set_fact:
        localssd_scsi_devices:
          stdout_lines: []
      when: localssd_scsi_devices|failed

    - name: Discover SSD device names (NVME)
      shell: |
        ls -1 /dev/disk/by-id/nvme-nvme.*
      register: localssd_nvme_devices

    - name: No NVME SSD devices
      set_fact:
        localssd_nvme_devices:
          stdout_lines: []
      when: localssd_nvme_devices|failed

    - name: Set local SSD variables
      set_fact:
        localssd_all_devices_count: '{{ localssd_nvme_devices.stdout_lines|length + localssd_scsi_devices.stdout_lines|length }}'
        localssd_all_devices: '{{ localssd_nvme_devices.stdout_lines + localssd_scsi_devices.stdout_lines }}'

    - name: Install prerequisites
      package:
        name: mdadm
        state: installed

    - name: Make RAID0 device
      shell: |
        mdadm --create /dev/md/google-local-ssd \
            --level=0 \
            --raid-devices='{{ localssd_all_devices_count }}' \
            {{ localssd_all_devices|join(" ") }}
      args:
        creates: '/dev/md/google-local-ssd'
      when: 'localssd_all_devices_count > 0'

    - name: Format RAID0 device
      filesystem:
        dev: '/dev/md/google-local-ssd'
        fstype: ext4
        opts: '-m 0 -L /srv/google-local-ssd'
      when: 'localssd_all_devices_count > 0'

    - name: Make mountpoint
      file:
        dest: '/srv/google-local-ssd'
        state: directory
      when: 'localssd_all_devices_count > 0'

    - name: Mount local SSD filesystem
      mount:
        path: '/srv/google-local-ssd'
        src: '/dev/md/google-local-ssd'
        state: mounted
        fstype: ext4
        opts: 'rw,nofail'
      when: 'localssd_all_devices_count > 0'
```

**Comment by riccardomurri** [comment]
To go on from here, I would need to know how you plan/want to use local SSDs (share them via NFS across the cluster? use as local cache directory on worker nodes? optimize for writes?)

**Comment by xiaoyanh** [comment]
Cool! Thank you very much!

I plan to use the SSD on the local worker nodes for fast reads. My understanding is rather vague here: Wouldn’t using them as NFS servers defeat the speed boost of having them locally?

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by riccardomurri** [comment]
hi @xiaoyanh 

> I plan to use the SSD on the local worker nodes for fast reads.

Makes total sense.  Google's "local SSD" disks have a fixed size (375 GB currently); if there is more than one, do you want them aggregated into a single large partition (current behavior of the playbook above) or do you prefer to have separate partitions?

> Wouldn’t using them as NFS servers defeat the speed boost of having them locally?

Yes, totally.  But it's still faster than serving NFS from spinning disks so it might have a use case.

**Comment by riccardomurri** [comment]
Fixed in #572

**Comment by riccardomurri** [comment]
Now that support for "local SSD" is merged into "master", you can create nodes with local SSD storage (mounted under `/srv/google-local-ssd`) with a config like this::

    [cluster/su16g]
    cloud=google
    setup=slurm+gpu

    login=google
    image_id=ubuntu-1604-xenial-v20171107b
    # note that shared-core instances (e.g., g1-small) do not support local SSD
    flavor=n1-standard-1
    security_group=default

    master_nodes=1
    worker_nodes=4

    [cluster/su16g/worker]
    # local SSD space of 2x375GB
    local_ssd_count=2
    local_ssd_interface=NVME
    scheduling=preemptible
